File: xuexi_py_backend/app/schemas/course.py
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


def to_camel(value: str) -> str:
    first, *rest = value.split("_")
    return first + "".join(part.capitalize() for part in rest)


class ApiModel(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)


class CourseRead(ApiModel):
    id: str
    title: str
    description: str


class CodeComparison(ApiModel):
    javascript: str
    python: str
    note: str


class ExplanationLine(ApiModel):
    code: str
    description: str


class CommonError(ApiModel):
    title: str
    description: str
    code: str | None = None


class RealWorldExample(ApiModel):
    title: str
    description: str
    code: str


class ExerciseRead(ApiModel):
    id: str
    type: Literal["fill", "choice", "predict", "debug", "code", "review", "rewrite", "incident", "design"]
    prompt: str
    code: str | None = None
    options: list[str]
    difficulty: Literal[1, 2, 3, 4, 5] = 1
    starter_code: str | None = None
    test_cases: list["ExerciseTestCase"] = Field(default_factory=list)
    hints: list[str] = Field(default_factory=list)


class ExerciseTestCase(ApiModel):
    name: str
    code: str


class LessonRead(ApiModel):
    id: str
    stage_id: str
    order: int
    title: str
    subtitle: str
    duration: int
    difficulty: Literal[1, 2, 3, 4, 5]
    importance: Literal["must", "frequent", "read", "skip"]
    status: Literal["available", "coming-soon"]
    one_liner: str
    comparison: CodeComparison
    explanation: list[ExplanationLine]
    common_errors: list[CommonError]
    real_world: RealWorldExample
    exercise: ExerciseRead
    exercises: list[ExerciseRead]
    simulated_output: str | None = None


class StageRead(ApiModel):
    id: str
    order: int
    title: str
    short_title: str
    description: str
    lesson_count: int
    status: Literal["active", "coming-soon"]
    lesson_ids: list[str]


class PracticeItemRead(ApiModel):
    id: str
    lesson_id: str
    title: str
    stage_id: str
    exercise: ExerciseRead


class SkillRead(ApiModel):
    id: str
    stage_id: str
    order: int
    title: str
    description: str
    level: Literal["foundation", "intermediate", "advanced", "expert"]
    mastery_threshold: int
    lesson_ids: list[str]
    prerequisite_ids: list[str]


class ProjectTaskRead(ApiModel):
    id: str
    order: int
    title: str
    description: str
    starter_code: str | None = None
    acceptance_criteria: list[str]
    solution_notes: str | None = None


class ProjectRead(ApiModel):
    id: str
    order: int
    title: str
    summary: str
    description: str
    difficulty: Literal[1, 2, 3, 4, 5]
    estimated_hours: int
    status: Literal["available", "coming-soon"]
    skill_ids: list[str]
    tasks: list[ProjectTaskRead]


class LabStepRead(ApiModel):
    id: str
    order: int
    title: str
    instructions: str
    commands: list[str]
    verification: list[str]
    hints: list[str]


class LabRead(ApiModel):
    id: str
    order: int
    title: str
    summary: str
    description: str
    level: Literal["advanced", "expert"]
    kind: Literal["engineering", "source", "performance", "incident", "architecture", "ai"]
    estimated_hours: int
    status: Literal["available", "coming-soon"]
    objectives: list[str]
    skill_ids: list[str]
    steps: list[LabStepRead]


class CourseCatalogRead(ApiModel):
    course: CourseRead
    stages: list[StageRead]
    lessons: list[LessonRead]
    practice_challenges: list[PracticeItemRead]
    skills: list[SkillRead]
    projects: list[ProjectRead]
    labs: list[LabRead]

File: xuexi_py_frontend/src/types/course.ts
export type Difficulty = 1 | 2 | 3 | 4 | 5
export type Importance = 'must' | 'frequent' | 'read' | 'skip'
export type LessonStatus = 'available' | 'coming-soon'

export interface ApiResponse<T> {
  code: number
  message: string
  data: T | null
}

export interface CodeComparison {
  javascript: string
  python: string
  note: string
}

export interface ExplanationLine {
  code: string
  description: string
}

export interface Exercise {
  id?: string
  type: 'fill' | 'choice' | 'predict' | 'debug' | 'code' | 'review' | 'incident' | 'design'
  prompt: string
  code?: string
  options: string[]
  answer: string
  explanation: string
  difficulty?: Difficulty
  starterCode?: string
  testCases?: { name: string; code: string }[]
  hints?: string[]
}

export interface Lesson {
  id: string
  stageId: string
  order: number
  title: string
  subtitle: string
  duration: number
  difficulty: Difficulty
  importance: Importance
  status: LessonStatus
  oneLiner: string
  comparison: CodeComparison
  explanation: ExplanationLine[]
  commonErrors: { title: string; description: string; code?: string }[]
  realWorld: { title: string; description: string; code: string }
  exercise: Exercise
  exercises?: Exercise[]
  simulatedOutput?: string
}

export interface Stage {
  id: string
  order: number
  title: string
  shortTitle: string
  description: string
  lessonCount: number
  status: 'active' | 'coming-soon'
  lessonIds: string[]
}

export interface PracticeItem {
  id: string
  lessonId: string
  title: string
  stageId: string
  exercise: Exercise
}

export interface CourseSummary {
  id: string
  title: string
  description: string
}

export interface CourseCatalog {
  course: CourseSummary
  stages: Stage[]
  lessons: Lesson[]
  practiceChallenges: PracticeItem[]
  skills: Skill[]
  projects: Project[]
  labs: LearningLab[]
}

export type SkillLevel = 'foundation' | 'intermediate' | 'advanced' | 'expert'

export interface Skill {
  id: string
  stageId: string
  order: number
  title: string
  description: string
  level: SkillLevel
  masteryThreshold: number
  lessonIds: string[]
  prerequisiteIds: string[]
}

export interface ProjectTask {
  id: string
  order: number
  title: string
  description: string
  starterCode?: string
  acceptanceCriteria: string[]
  solutionNotes?: string
}

export interface Project {
  id: string
  order: number
  title: string
  summary: string
  description: string
  difficulty: Difficulty
  estimatedHours: number
  status: LessonStatus
  skillIds: string[]
  tasks: ProjectTask[]
}

export type LabLevel = 'advanced' | 'expert'
export type LabKind = 'engineering' | 'source' | 'performance' | 'incident' | 'architecture' | 'ai'

export interface LabStep {
  id: string
  order: number
  title: string
  instructions: string
  commands: string[]
  verification: string[]
  hints: string[]
}

export interface LearningLab {
  id: string
  order: number
  title: string
  summary: string
  description: string
  level: LabLevel
  kind: LabKind
  estimatedHours: number
  status: LessonStatus
  objectives: string[]
  skillIds: string[]
  steps: LabStep[]
}

export interface ExerciseResult {
  lessonId: string
  attempts: number
  correctAttempts: number
  lastAttemptAt: string
}

export interface ProjectTaskSubmission {
  notes: string
  checkedCriteria: number[]
  submittedAt: string
}

export interface LabStepSubmission {
  notes: string
  checkedVerification: number[]
  completedAt: string
}

export interface LearningState {
  completed: string[]
  bookmarked: string[]
  history: { lessonId: string; completedAt: string }[]
  wrongLessonIds: string[]
  exerciseAttempts: number
  exerciseCorrect: number
  exerciseResults: Record<string, ExerciseResult>
  projectSubmissions: Record<string, ProjectTaskSubmission>
  labSubmissions: Record<string, LabStepSubmission>
}
